### PR TITLE
WINC-738: Define Kubelet service in services ConfigMap

### PR DESCRIPTION
**Pull request by sebsoto**
This PR moves the responsibility of managing the kubelet service into WICD. This required WMCO to pick up the responsibility of getting and parsing the ignition file for kubelet arguments. With this commit, WICD will reconcile an incorrectly configured kubelet service on Windows nodes.

### CODE REVIEW

File: cmd/operator/main.go
Comment (aravindhp) [inline cmd/operator/main.go:181]: We should call out somewhere that we are changing the behavior when compared to WMCB when we were doing it once per node.

File: pkg/ignition/ignition.go
Comment (aravindhp) [inline pkg/ignition/ignition.go:44]: `machineConfigs`?

File: pkg/ignition/ignition.go
Comment (aravindhp) [inline pkg/ignition/ignition.go:52]: Is there a way to pick the latest? Or add a doc here saying why we are picking the first one we find.

File: pkg/ignition/ignition.go
Comment (aravindhp) [inline pkg/ignition/ignition.go:79]: There is an opportunity to make this more oops like. Can you take a stab?

File: pkg/ignition/ignition.go
Comment (aravindhp) [inline pkg/ignition/ignition.go:116]: Please add some docs around what you are doing here

File: pkg/ignition/ignition.go
Comment (aravindhp) [inline pkg/ignition/ignition.go:107]: Shouldn't we also include these?

**Review by aravindhp — CHANGES_REQUESTED** [review]
Thanks for working on this, @sebsoto 

>This commit moves the responsibility of managing the kubelet service into
WICD. This required WMCO to pick up the responsibility of getting and
parsing the ignition file for kubelet arguments.

Is the above accurate? WMCO is still actually managing the kubelet. It is the follow up PR that will move the responsibility. You are just laying the groundwork.

>With this commit, WICD will reconcile an incorrectly configured kubelet service on Windows nodes.

I find the above statement confusing. We are not going to fix an incorrect kubelet service once we have already prepped the node.

File: pkg/daemon/controller/controller.go
Comment (aravindhp) [inline pkg/daemon/controller/controller.go:357]: Please mention that this goes against the WICD design principle add a link to a Jira story where you will address this.

File: pkg/ignition/ignition.go
Comment (sebsoto) [reply pkg/ignition/ignition.go:109]: Added comment explaining why thats not needed

File: cmd/operator/main.go
Comment (sebsoto) [reply cmd/operator/main.go]: Changed this as it was unnecessary.

File: pkg/ignition/ignition.go
Comment (sebsoto) [reply pkg/ignition/ignition.go]: Picking the latest now

File: pkg/ignition/ignition.go
Comment (saifshaikh48) [inline pkg/ignition/ignition.go:89]: can be moved closer to where it is used

File: bundle/manifests/windows-machine-config-operator.clusterserviceversion.yaml
Comment (saifshaikh48) [inline bundle/manifests/windows-machine-config-operator.clusterserviceversion.yaml:327]: Curious if you dug into [why watch is needed](https://github.com/openshift/windows-machine-config-operator/pull/1235#discussion_r964225221)?

File: pkg/ignition/ignition.go
Comment (saifshaikh48) [inline pkg/ignition/ignition.go:79]: can delete these lines

File: pkg/ignition/ignition.go
Comment (saifshaikh48) [inline pkg/ignition/ignition.go:52]: do we need this log?

**Review by saifshaikh48 — COMMENTED** [review]
Nice splitting of the work and looks good overall!

I assume the followup PR to create all the kubelet-required files from WMCO will churn some of this code when trying to avoid getting and parsing ignition every time we reconcile a node.

**Review by aravindhp — CHANGES_REQUESTED** [review]

**Comment by aravindhp** [comment]
> I assume the followup PR to create all the kubelet-required files from WMCO will churn some of this code when trying to avoid getting and parsing ignition every time we reconcile a node.

Really? I thought @sebsoto wanted to retain the WMCB behavior of getting and parsing ignition every a node is reconciled.

File: pkg/ignition/ignition.go
Comment (sebsoto) [reply pkg/ignition/ignition.go:50]: I wasn't able to find a way to do that with a name field selector, I'm only aware of exact matching for a field.
Additionally, theres no annotation or label which denotes it being a rendered worker type

File: pkg/ignition/ignition.go
Comment (sebsoto) [reply pkg/ignition/ignition.go:52]: I'm not sure the value of this considering this would only be used when bootstrapping, its a lot more valuable for linux nodes. I've added [another story](https://issues.redhat.com/browse/WINC-897) in a TODO which is semi-related.

File: pkg/services/services.go
Comment (sebsoto) [reply pkg/services/services.go:177]: Made this into a function

File: controllers/configmap_controller.go
Comment (sebsoto) [reply controllers/configmap_controller.go:398]: Yes, that is me thinking too far ahead. I've changed the predicate to filter out all MachineConfig events, leaving it in the `Watches` is important as it allows the cached client to pull the MachineConfig object. Added a [TODO](https://issues.redhat.com/browse/WINC-897) to stop filtering events and act upon changes.

Comment (aravindhp) [reply controllers/configmap_controller.go:397]: > leaving it in the Watches is important as it allows the cached client to pull the MachineConfig object

That sounds really weird. This shouldn't be the reason to Watch the object. 

>Added a [TODO](https://issues.redhat.com/browse/WINC-897) to stop filtering events and act upon changes.

This sounds like a risky TODO. Do we know how often these rendered config changes? We don't want a situation where we are unnecessarily re-configuring Windows nodes.

File: pkg/ignition/ignition.go
Comment (aravindhp) [reply pkg/ignition/ignition.go]: The value would be when debugging a customer issue. We can easily look at the node object and figure out where we got the kubelet args from.

File: pkg/daemon/controller/controller.go
Comment (saifshaikh48) [inline pkg/daemon/controller/controller.go:372]: Shouldn't we also compare the address listed in the instances ConfigMap instead? Or am I missing something that makes this check unnecessary

`if instance.Address == addr.String() || instance.IPv4Address == addr.String() {`

Comment (saifshaikh48) [reply pkg/daemon/controller/controller.go:373]: I suspect this is why the platform-none test is failing with `unable to find IPv4 address to use as Node IP for this instance` -- there is no guarantee the interface addresses for a specific interface will contain an IPv4 address, right?

Comment (saifshaikh48) [reply pkg/daemon/controller/controller.go:373]: It looks like your latest debug commit makes it clear. `addr.String()` returns a value with subnet appended

```
I0928 22:14:47.569874    4660 controller.go:372] Checking if 192.168.59.27 == fe80::f127:441b:58:4b46/64
I0928 22:14:47.569874    4660 controller.go:372] Checking if 192.168.59.27 == 192.168.59.27/27
I0928 22:14:47.569874    4660 controller.go:372] Checking if 192.168.59.27 == ::1/128
I0928 22:14:47.569874    4660 controller.go:372] Checking if 192.168.59.27 == 127.0.0.1/8
I0928 22:14:47.569874    4660 controller.go:372] Checking if 192.168.59.27 == fe80::d156:975f:69ef:f1c4/64
I0928 22:14:47.569874    4660 controller.go:372] Checking if 192.168.59.27 == 172.27.0.1/20
E0928 22:14:47.569874    4660 controller.go:252] unable to find IPv4 address to use as Node IP for this instance
```

Comment (sebsoto) [reply pkg/daemon/controller/controller.go:373]: The test was failing because the local address wasn't being converted into an ipv4 address before the comparison.
String-ing a net.Addr will result in an address like `192.168.0.1/32 `, stringing a net.IP will give `192.168.0.1`

Comment (sebsoto) [reply pkg/daemon/controller/controller.go:373]: > there is no guarantee the list of all interface addresses on a specific instance will contain an IPv4 address, right?

As part of `GetInstances()` .IPv4Address needs to be populated or else the CM parse will fail.

Comment (saifshaikh48) [reply pkg/daemon/controller/controller.go]: I guess my question is related to the instance -- will the addr list returned by `LocalInterfaceAddresses()` always contain an IPv4 value? If not, then IMO we should also compare with `instance.Address`

File: controllers/configmap_controller.go
Comment (mansikulkarni96) [inline controllers/configmap_controller.go:415]: I suggest we follow the new gofmt conventions keeping non-code blocks unindented.

File: pkg/services/services.go
Comment (mansikulkarni96) [inline pkg/services/services.go:22]: space needed after //

File: pkg/daemon/controller/controller.go
Comment (mansikulkarni96) [inline pkg/daemon/controller/controller.go:357]: same applies here for unindented comments.

File: pkg/daemon/controller/controller.go
Comment (sebsoto) [reply pkg/daemon/controller/controller.go]: It would as per the requirements of WMCO. We don't support IPv6.

File: controllers/configmap_controller.go
Comment (saifshaikh48) [reply controllers/configmap_controller.go:406]: +1 on removing this watch and `renderedWorkerPredicate`. You can put the TODO with issue link here, and if you'd like update the issue engineering details to capture the watch + predicate.

File: test/e2e/create_test.go
Comment (saifshaikh48) [inline test/e2e/create_test.go:212]: Should we append the address to the job name instead of just `print-logs-byoh`?

**Review by saifshaikh48 — COMMENTED** [review]
Nice job getting this working. Just a couple comments

File: controllers/configmap_controller.go
Comment (aravindhp) [reply controllers/configmap_controller.go:406]: I continue to find this strange. It is one thing to say we need watch for changes in the future but another to add it here because it allows the cached client to pull the MachineConfig object without knowing why. Are you sure you are not missing adding it to a scheme somewhere in the code?

File: pkg/ignition/ignition.go
Comment (aravindhp) [inline pkg/ignition/ignition.go:32]: Nice work

File: pkg/ignition/ignition.go
Comment (aravindhp) [inline pkg/ignition/ignition.go:95]: Please add an info message here saying which MC we picked. This will help debugging customer issues.

File: pkg/services/services.go
Comment (aravindhp) [inline pkg/services/services.go:93]: Thanks for modularizing this but my question still remains. Why do you have to add these args in two places?

File: test/e2e/create_test.go
Comment (aravindhp) [reply test/e2e/create_test.go:212]: +1

**Review by aravindhp — CHANGES_REQUESTED** [review]

File: bundle/manifests/windows-machine-config-operator.clusterserviceversion.yaml
Comment (sebsoto) [reply bundle/manifests/windows-machine-config-operator.clusterserviceversion.yaml]: I was able to remove this. not sure why you were seeing this

File: controllers/configmap_controller.go
Comment (aravindhp) [inline controllers/configmap_controller.go:408]: This function can be removed now.

File: test/e2e/create_test.go
Comment (aravindhp) [inline test/e2e/create_test.go:173]: I suggest making this a helper function.

**Review by aravindhp — APPROVED** [review]

File: controllers/configmap_controller.go
Comment (saifshaikh48) [reply controllers/configmap_controller.go:408]: +1
the `TODO` can be moved to `SetupWithManager` (line 403)

File: bundle/manifests/windows-machine-config-operator.clusterserviceversion.yaml
Comment (saifshaikh48) [reply bundle/manifests/windows-machine-config-operator.clusterserviceversion.yaml]: Thanks for investigating

**Review by saifshaikh48 — COMMENTED** [review]
/lgtm

**Review by saifshaikh48 — COMMENTED** [review]
/lgtm

**Comment by aravindhp** [comment]
Blocked by OCPBUGS-2175

File: pkg/services/services.go
Comment (jrvaldes) [inline pkg/services/services.go:128]: @sebsoto, with the need for the hostname override in GCP, I feel there is a disconnection here between `getKubeletHostnameOverride()`  and the way the variable is getting resolved.  `Get-EC2InstanceMetadata` is platform specific and will only work in AWS. 

Can this be done in a way that could be extended based on the platform type?

Comment (jrvaldes) [reply pkg/services/services.go:128]: The `Get-EC2InstanceMetadata` is part of the [AWS.Tools for PowerShell](https://docs.aws.amazon.com/powershell/latest/userguide/pstools-welcome.html), and it comes pre-installed on Windows machines. What about BYOH instances in AWS, will the AWS.Tools for PowerShell be available?

File: pkg/services/services.go
Comment (sebsoto) [reply pkg/services/services.go:21]: These are not, they are specific to klog

**Comment by sebsoto** [comment]
@jrvaldes considering the length of time this PR was approved and LGTM'd before you reviewed, I'd prefer to skip the comments you've left, if you don't consider them blockers for this PR.

**Comment by jrvaldes** [comment]
> @jrvaldes considering the length of time this PR was approved and LGTM'd before you reviewed, I'd prefer to skip the comments you've left, if you don't consider them blockers for this PR.

Agree. Most of the comments will be addressed in the follow-up PRs that are based on these changes.

**Comment by aravindhp** [comment]
@sebsoto just be aware that the PR will merge even if `platform-none-vsphere-e2e-operator` fails as it is still marked as optional. You would need the hold to ensure it passes.

**Comment by sebsoto** [comment]
Platform none test passed, removing hold
/hold cancel

**Comment by aravindhp** [comment]
@sebsoto I am hoping [this](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_windows-machine-config-operator/1260/pull-ci-openshift-windows-machine-config-operator-master-azure-e2e-ccm-install/1585695398150803456) is a flake:

```
2022/10/27 20:35:47 Unable to get logs associated with pod: job-name=print-logs-e2e-wm-5xh9r-job-gbpbj
2022/10/27 20:35:47 failed to collect logs from e2e-wm-5xh9r: error waiting for job to succeed:
```

**Comment by sebsoto** [comment]
@aravindhp That job doesn't make any sense to me. I'm not sure why its appearing on this PR. Its running using test code from another branch

**Comment by aravindhp** [comment]
> @aravindhp That job doesn't make any sense to me. I'm not sure why its appearing on this PR. Its running using test code from another branch

Whoa! How is that even possible? Please check with @Fedosin who I think added the job.

**Pull request merged**
